### PR TITLE
add cert refresh time and unit test for query

### PR DESCRIPTION
**Pull request by gunsluo**

### CODE REVIEW

**Comment by gunsluo** [comment]
@clearcodecn  please review and merge.

**Comment by gunsluo** [comment]
@clearcodecn   rebase master.

please review and merge.

File: client.go [hunk 1]
```diff
@@ -339,3 +341,41 @@ func (c *client) onceDownloadCertificates(ctx context.Context) error {
...
 }
+
+type secrets struct {
+	mutex    sync.RWMutex
+	deadline time.Time
+	all      map[string]*rsa.PublicKey
+}
+
+func (s *secrets) isUpgrade() bool {
+	if s.deadline.Before(time.Now()) {
```
Comment (clearcodecn) [inline client.go:352]: `deadline` also should be protected by `sync.RWMutex`

Comment (clearcodecn) [inline client.go:352]: `deadline` also should be protected by `sync.RWMutex`

**Review by clearcodecn — COMMENTED** [review]
LGTM

**Pull request merged**